precovery/brute_force.py

```diff
@@ -0,0 +1,387 @@
+import os
+
+from sqlalchemy import all_
+
+from precovery.precovery_db import PrecoveryDatabase
+
+os.environ["OMP_NUM_THREADS"] = "1"
+os.environ["OPENBLAS_NUM_THREADS"] = "1"
+os.environ["MKL_NUM_THREADS"] = "1"
+os.environ["VECLIB_MAXIMUM_THREADS"] = "1"
+os.environ["NUMEXPR_NUM_THREADS"] = "1"
+
+import logging
+import multiprocessing as mp
+import time
+from functools import partial
+from typing import Iterable, List, Optional, Tuple
+
+import numpy as np
+import pandas as pd
+from astropy.time import Time
+from sklearn.neighbors import BallTree
+
+from .healpix_geom import radec_to_healpixel, radec_to_healpixel_with_neighbors
+
+# replace this usage with Orbit.compute_ephemeris
+from .orbit import Orbit
+from .residuals import calcResiduals
+from .utils import _checkParallel, _initWorker, calcChunkSize, yieldChunks
+
+logger = logging.getLogger(__name__)
+
+__all__ = [
+    "get_observations_and_ephemerides_for_orbits",
+    "attribution_worker",
+    "attributeObservations",
+]
+
+
+def get_observations_and_ephemerides_for_orbits(
+    orbits: Iterable[Orbit],
+    mjd_start: float,
+    mjd_end: float,
+    precovery_db: PrecoveryDatabase,
+    obscode: str = "W84",
+):
+    """
+    Returns:
+    - ephemerides for the orbit list, propagated to all epochs for frames in 
+    the range (mjd_start, mjd_end)
+    - observations in the indexed PrecoveryDatabase consistent with intersecting
+    and neighboring frames for orbits propagated to each of the represented epochs
+
+    ***Currently breaks on non-NSC datasets
+        TODO propagation targets should be unique frames wrt obscode, mjd
+    """
+
+    # Find all the mjd we need to propagate to
+    all_frame_mjd = precovery_db.frames.idx.unique_frame_times()
+    frame_mjd_within_range = [
+        x for x in all_frame_mjd if (x > mjd_start and x < mjd_end)
+    ]
+
+    ephemeris_dfs = []
+    frame_dfs = []
+    # this should pass through obscode...rather should use the relevant frames' obs codes
+    for orbit in orbits:
+        eph = orbit.compute_ephemeris(obscode=obscode, epochs=frame_mjd_within_range)
+        mjd = [w.mjd for w in eph]
+        ra = [w.ra for w in eph]
+        dec = [w.dec for w in eph]
+        ephemeris_df = pd.DataFrame.from_dict(
+            {
+                "mjd_utc": mjd,
+                "RA_deg": ra,
+                "Dec_deg": dec,
+            }
+        )
+        ephemeris_df["orbit_id"] = orbit.orbit_id
+        ephemeris_df["observatory_code"] = obscode
+        ephemeris_dfs.append(ephemeris_df)
+
+        # Now we gathetr the healpixels and neighboring pixels for each propagated position
+        healpix = radec_to_healpixel_with_neighbors(
+            ra, dec, precovery_db.frames.healpix_nside
+        )
+
+        frame_df = pd.concat(
+            [
+                pd.DataFrame.from_dict(
+                    {
+                        "mjd_utc": [x[0] for y in range(9)],
```

Comment:
What is `range(9)` doing here exactly? Might be worth adding a comment since it's a hard-coded list length.

Reply:
Good catch, this is unclear. Healpixel plus neighbors is a set of 9 healpixels